### PR TITLE
v0.2.3: calibration-aware circuit packing, QKAN-aware optimizers, and release tooling

### PR DESCRIPTION
**Pull request by Jim137**
Promotes the `v0.2.3dev` line to master: a calibration-aware circuit-packing toolkit with hardware-verified CUDA-Q and qiskit paths, QKAN-aware optimizers and CUDA-graph training capture, multi-architecture CuTe wheels, an agent onboarding skill, and release tooling. 31 commits, 67 files, +11,845/−568.

## Circuit packing & calibration-aware layouts (new)

- `qkan.solver.layout`: provider-neutral `DeviceProfile` (from qiskit backends or Braket devices) with `rank_qubits`, `best_subgraph`, and `tile_disjoint` — calibration-scored qubit/subgraph/tile selection with dead-edge pruning and threshold/strict semantics.
- `qkan.solver.packing`: `pack_circuit` packs `k` copies of a circuit onto disjoint calibrated tiles through one entry point for both stacks — qiskit `QuantumCircuit` (SWAP-free transpilation pinned to tiles) and plain `@cudaq.kernel` (gates extracted from compiled Quake IR and rebuilt at physical indices). One `PackedCircuit` type serves both: cross-stack `observable()` mapping, counts readout via `expectation`/`basis_kernel`, per-copy parameter batches / `block_args_batch` with `rebind`, and pooled `"mean"` readout for shot reduction.
- Verified on real hardware (Rigetti via Braket): packed multi-qubit jobs and the observe route return program-indexed, logically correct results; vendor rewiring behavior and per-term task costs are documented in the packing guide.
- New test suites under `skills/tests/`: `test_packing.py` + `test_packing_cudaq.py` (28 tests, qiskit/cudaq importorskip; no top-level tests/ directory).

## Hardware solvers

- qiskit: calibration thresholds for `best_qubits`, `initial_layout="auto"` for packed jobs, Runtime job accounting via `job_event_callback`.
- CUDA-Q: calibration-ranked `initial_layout` on the packed path (zero-angle padding defeats compiler renumbering), Quantum Machines (Qoperator) target support.

## Training & inference performance

- QKAN-aware optimizers: `QKANAdamMini`, `QKANBeliefMini`/`TritonAdaBelief` (BF16 mini-state), `QKANSpectralMini`, `QKANMuon` (incl. DDP-sharded Newton–Schulz), L-BFGS finisher schedules, plus Hessian analysis tooling.
- Rep-loop activation checkpointing, fused Triton epilogue (fwd+bwd), CUDA-graph capture for inference and full train steps (`compile_inference`, `graph_submodules`, `make_graphed_train_step`).
- Backend-matched activations aligned with `torch.nn.Linear`, fp8 components, `p_dim` 2D-flat parameter storage, hot-path micro-optimizations.
- Solver architecture: `QKANSolver` ABC, backends grouped into subdirectories, unified naming, CUDA-family-split extras.

## Build & distribution

- Wheel index split by CUDA family (cu12/cu13) with matching extras (`qkan[cute-cu12]`, …).
- Multi-architecture CuTe wheels (sm80/90/100/120 + PTX) with a runtime GPU-compatibility probe: incompatible prebuilt kernels warn and fall back to JIT rebuild instead of failing at first launch.

## Tooling & docs

- `skills/qkan-guide`: agent onboarding skill (install/train/solvers/precision/packing/QPU routing) with a 10-case eval set — verified 7/7 positive correctness and 10/10 discoverability by independent agent runs (`skills/qkan-guide/BENCHMARK.md`).
- Release tooling: dev versions carry the source commit (`0.2.3dev+g<hash>`, composing with CUDA wheel tags; never on sdists or releases), and `release-prep.yml` automates the branch-then-tag flow — master never leaves dev state, tags live on disposable `release/vX.Y.Z` branches, version input is injection-safe via env binding.
- New/updated guides: packing, optimizers, inference/training graphs, applications; solver guide extended with real-device deployment, qubit-calibration layout, and error mitigation.

### CODE REVIEW

File: src/qkan/qkan.py [hunk 1]
```diff
@@ -538,38 +690,92 @@ def forward(self, x: torch.Tensor):
...
+            x,
+            self._theta_natural_v,
+            self._preacts_w_natural_v,
+            self._preacts_b_natural_v,
+        )
+        pw_oi = self._pw_oi
+        bw_oi = self._bw_oi
+        if self._mask_is_identity:
+            eff_pw = pw_oi
+            base_w = bw_oi
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/qkan.py:766]: Confirmed and fixed in b680007 — `_load_from_state_dict` now recomputes `_mask_is_identity` (and the new `_pb_is_nonzero`) from the loaded tensors, so pruned checkpoints keep their masks in both train and eval routes. Regression test: `skills/tests/test_review_regressions.py::test_pruned_mask_survives_state_dict_load` (covers the reverse direction too).

File: src/qkan/qkan.py [hunk 2]
```diff
@@ -1567,9 +1817,29 @@ def _closure():
...
+                        # trailing (reps+1, k) axes to per-edge magnitude.
+                        edge_mag = torch.sum(
+                            torch.abs(layer._theta_natural()), dim=(-2, -1)
+                        )
+                        kill = edge_mag < prune_threshold
+                        if not bool(kill.any()):
+                            continue
+                        mask_2d = layer._as_oi(layer.mask).data
+                        if kill.shape != mask_2d.shape:
+                            kill = kill.reshape(mask_2d.shape)
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/qkan.py:1869]: Confirmed and fixed in b680007 — the auto-prune loop now skips layers whose `kill.numel() != mask.numel()` (grouped/shared theta has no per-edge magnitude), documented in the `prune_every` docstring. Test: `test_grouped_layer_autoprune_does_not_raise`.

File: src/qkan/inference.py [hunk 3]
```diff
@@ -128,6 +129,186 @@ def replay(x: torch.Tensor) -> torch.Tensor:
...
+    The optimizer's ``param_groups`` are rewired to the new Parameter objects;
+    any existing ``optimizer.state`` is re-keyed best-effort (state tensors
+    survive but are NOT migrated to a new stream — re-init momentum state if
+    you saw any training before).
+    """
+    old_to_new: dict[int, torch.nn.Parameter] = {}
+    for name, p in list(model.named_parameters()):
+        if not p.requires_grad:
+            continue
+        new_p = torch.nn.Parameter(p.data.clone(), requires_grad=True)
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/inference.py:159]: Confirmed and fixed in b680007 — `_reallocate_params_on_stream` now migrates the id-keyed `_param_names` map (popping old ids so recycled ids can't attach wrong names) and clears `QKANMuon._partition_cache` so partitions rebuild against the live Parameters. Test: `test_graphed_param_swap_migrates_optimizer_names`.

File: src/qkan/optim/lbfgs.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        """Run one optimization step. ``closure`` is required.
+
+        L-BFGS *requires* a closure that zeros grads, runs forward+backward,
+        and returns the loss tensor. The early optimizer is also driven via
+        the same closure so user code stays uniform across the swap.
+        """
+        opt = self.current
+        # Both LBFGS and Adam-family accept tensor-returning closures at
+        # runtime; the typed stubs declare ``Callable[[], float]`` only.
+        loss = opt.step(closure)  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/optim/lbfgs.py:149]: Confirmed and fixed in b680007 — `QKAN.train_` now treats `LBFGSFinisher` like `torch.optim.LBFGS` in both the auto-closure and dispatch checks, so `adam_then_lbfgs(...)` runs through the closure branch in both phases. Test: `test_adam_then_lbfgs_through_train_` (crosses the phase switch).

File: src/qkan/qkan.py [hunk 5]
```diff
@@ -356,23 +562,45 @@ def to(self, *args, **kwargs):
             self.device = device
             for param in self.parameters():
                 param.data = param.to(device)
-        return super(QKANLayer, self).to(*args, **kwargs)
+        ret = super(QKANLayer, self).to(*args, **kwargs)
+        # Param storage may have moved — rebuild cached views over the new
+        # storage. No-op under p_dim=4.
+        self._init_view_caches()
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/qkan.py:583]: Confirmed and fixed in b680007 — `QKANLayer.to`/`QKAN.to` now honor `dtype=` keyword moves and `to()` refreshes `_needs_dtype_cast` after any dtype change. Test: `test_dtype_cast_flag_refreshes_on_to`.

File: src/qkan/qkan.py [hunk 6]
```diff
@@ -538,38 +690,92 @@ def forward(self, x: torch.Tensor):
...
+        if self._use_fused_epilogue and postacts.is_cuda:
+            pb = self._pb_oi if self.postact_bias_trainable else None
+            return QKANEpilogue.apply(postacts, eff_pw, pb, base_input, base_w)
+        if self.postact_bias_trainable:
+            main = torch.sum((postacts + self._pb_oi) * eff_pw, dim=2)
         else:
-            eff_pw = self.postact_weights * self.mask
-            main = torch.sum((postacts + self.postact_bias) * eff_pw, dim=2)
-            base = F.linear(base_input, self.base_weight * self.mask)
+            main = (postacts * eff_pw).sum(dim=2)
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/qkan.py]: Confirmed and fixed in b680007 — a cached `_pb_is_nonzero` flag (refreshed on checkpoint load and reset) keys the bias-skip optimization, so a frozen-but-nonzero `postact_bias` is honored on the eager, fused, and eval-fold paths without a per-forward device sync. Test: `test_frozen_nonzero_postact_bias_is_honored`.

File: README.md [hunk 7]
```diff
@@ -68,8 +68,11 @@ pip install qkan[gpu]
 To use the CuTe DSL solver (`solver="cute"`) with pre-built CUDA kernels:
 
 ```bash
-# Pre-built wheel (recommended — no compilation needed)
-pip install qkan[cute] --extra-index-url https://qkan.jimq.cc/whl/
+# Pre-built wheel (recommended — no compilation needed). Pick the CUDA
+# family matching the torch you already have installed:
+pip install qkan[cute] --extra-index-url https://qkan.jimq.cc/whl/cu12/
```
[the remark this answers is not in the document]
Comment (Jim137) [reply README.md]: Confirmed and fixed in b680007 — README's cu12 line now installs `qkan[cute-cu12]` (the bare `cute` extra aliases the cu13 runtime wheels); the qkan-guide skill install text got the same correction.

File: src/qkan/optim/triton_adabelief.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                        lr_eff=lr_eff,
+                        eps_sb=eps_sb,
+                        wd_factor=wd_factor,
+                    )
+                else:
+                    adabelief_step_(
+                        p,
+                        p.grad,
+                        m,
+                        s,
```
[the remark this answers is not in the document]
Comment (Jim137) [reply src/qkan/optim/triton_adabelief.py:246]: Confirmed and fixed in b680007 — `adabelief_step_` casts the gradient to the state dtype for the `lerp_` EMA update (the same pattern QKANBeliefMini uses), fixing the eager fallback for both bf16-state/fp32-param and the reverse. Test: `test_adabelief_eager_fallback_mixed_state_dtype`.
